### ISSUE TITLE
How to create 3D animation?

### ISSUE THREAD

**Issue by rkakash59** [UniFlowMatch/UFM#6]
In your project page, you show 3D animation generated using source and target images.

How can I create my own?

Will you please point out the code for it.

**Comment by rkakash59** [comment]
I have written this code for generating 3D animation.

`def make_ufm_3d_animation(
    warped_image: np.ndarray,
    flow: np.ndarray,
    output_gif_path: str,
    num_steps: int = 20,
    fps: int = 10
):
    flow = flow.transpose(1, 2, 0)
    h, w = flow.shape[:2]
    # Generate pixel coordinate grid
    xx, yy = np.meshgrid(np.arange(w), np.arange(h), indexing='xy')
    base_map = np.stack([xx, yy], axis=-1)

    frames = []

    for alpha in np.linspace(0, 1, num_steps):
        # Compute intermediate flow
        interpolated_flow = flow * alpha


        # Separate x and y channels
        map_x = base_map[..., 0] - interpolated_flow[..., 0]
        map_y = base_map[..., 1] - interpolated_flow[..., 1]
        map_x = map_x.astype(np.float32)
        map_y = map_y.astype(np.float32)

        # Remap using current flow
        moved = cv2.remap(
            warped_image, map_x, map_y,
            interpolation=cv2.INTER_LINEAR,
            borderMode=cv2.BORDER_CONSTANT,
            borderValue=(255, 255, 255)
        )

        # # Convert to RGB for imageio
        # moved_rgb = cv2.cvtColor(moved, cv2.COLOR_BGR2RGB)
        frames.append(moved)

    # Write out GIF
    Path(output_gif_path).parent.mkdir(parents=True, exist_ok=True)
    imageio.mimsave(output_gif_path, frames, fps=fps)
    print(f"Saved 3D‐motion GIF to {output_gif_path}")`

I am taking the warped source image and applying the predicted flow stepwise to have the warped image in target image pose. The problem I observe is there is  major deformation in some areas of the image. Is it because the flow is not quite accurate or is there something I am missing ?

**Comment by infinity1096** [comment]
Hi, thanks for trying out the visualization.

First, I’d like to clarify that the UFM visualization is purely 2D—it only appears 3D due to the way the flow is interpolated.

Second, the visualization is not actually performing warping. In warping, you bring the color from the target location back to the source, meaning the colors can change due to interpolation at the target. What we show instead is splatting: we take the color at the source image and move it to an interpolated position based on the predicted flow. This means the colors remain unchanged—only their positions are shifted.

Try this code:

```
def paint_color_to_location(source_image, source_mask, target_image, flow) -> np.ndarray:
    """
    Paint the source image into the target image, according to flow.

    Args:
    source_image: np.ndarray of shape (H, W, 3), normalized to [0, 1]
    target_image: np.ndarray of shape (H, W, 3), normalized to [0, 1]
    flow: np.ndarray of shape (H, W, 2)
    source_mask: covisible mask represented in source image.

    Returns:
    warped_image: target_image warped according to flow into frame of source image
    np.ndarray of shape (H, W, 3), normalized to [0, 1]

    """
    # assert source_image.shape[-1] == 3
    # assert target_image.shape[-1] == 3

    source_image = source_image.copy()
    target_image = target_image.copy()
    source_mask = source_mask.copy()
    flow = flow.copy()

    assert flow.shape[-1] == 2

    # Get the shape of the source image
    height, width = source_image.shape[:2]
    target_height, target_width = target_image.shape[:2]

    # Create mesh grid
    x, y = np.meshgrid(np.arange(width), np.arange(height))

    # Apply flow displacements
    flow_x, flow_y = flow[..., 0], flow[..., 1]
    x_new = x + flow_x + 0.5
    y_new = y + flow_y + 0.5

    # color to this location
    grid_x = np.floor(x_new).astype(np.int32)
    grid_y = np.floor(y_new).astype(np.int32)

    within_bounds = (grid_x >= 0) & (grid_x < target_width) & (grid_y >= 0) & (grid_y < target_height)

    valid_grid_x = torch.from_numpy(grid_x[within_bounds])
    valid_grid_y = torch.from_numpy(grid_y[within_bounds])
    valid_colors = torch.from_numpy(source_image[within_bounds]).float()

    painted_image_color = torch.zeros_like(torch.from_numpy(target_image), dtype=torch.float32)
    painted_image_weight = torch.zeros((target_height, target_width), dtype=torch.float32)

    torch.index_put_(
        painted_image_color,
        (valid_grid_y, valid_grid_x),
        valid_colors,
        accumulate=True
    )

    torch.index_put_(
        painted_image_weight,
        (valid_grid_y, valid_grid_x),
        torch.from_numpy(source_mask[within_bounds]),
        accumulate=True
    )

    average_color = painted_image_color / (painted_image_weight[..., None] + 1e-8)
    white_image = np.ones_like(target_image) * 255.0

    source_mask[~within_bounds] = 0.0

    painted_image_weight = torch.clip(painted_image_weight, 0, 1)

    painted_image = painted_image_weight[..., None] * average_color.numpy() + (1 - painted_image_weight[..., None]) * white_image

    return painted_image```

**Comment by rkakash59** [comment]
Hi, Thank you for me helping me out.

I am able to create the animation you show in your project page on my data.
